### ISSUE TITLE
tf.gather behavior depends on dtype

### ISSUE THREAD

**Issue by FordUniver** [tensorflow/tensorflow#45266]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): sort of
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.2
- Python version: 3.8
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: -
- GPU model and memory: -

**Describe the current behavior**
```
a = tf.ones((2, 3), dtype=tf.int32)
b = tf.constant([0, 1])
print(  tf.gather(a, b, batch_dims=-1).shape )
```
returns a shape of (2, ) whereas

```
a = tf.ones((2, 3), dtype=tf.int64)
b = tf.constant([0, 1])
print(  tf.gather(a, b, batch_dims=-1).shape )
```
returns a shape of (2, 2).

**Describe the expected behavior**
I'm not 100% which should be the expected behavior of tf.gather, but it seems like it should not depend on the dtype of a.

**Standalone code to reproduce the issue**
See above

**Other info / logs** Include any logs or source code that would be helpful to
–

**Comment by amahendrakar** [comment]
@FordUniver,
I did not observe any difference on running the code with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/a23572d7bc9e921b2a9ad02b1f82287c/45266-2-2.ipynb#scrollTo=ilg8TJQUKorm) and [TF v2.3](https://colab.research.google.com/gist/amahendrakar/f0893034cd88e71c350ed8b6c3d4d647/45266.ipynb). With both the versions the shape is `(2, 2)`. Please check the attached gist for reference. 

Also, could you please try running the code in a virtual environment with TF v2.3 and check if you are facing the same issue? Thanks!

**Comment by FordUniver** [comment]
After testing a bit, it seems like this only occurred for me in one particular, active kernel running in a particular conda environment and that the behavior changed after I restarted the kernel. I'm not sure what caused it in the first place, since the code snippet I posted reloads TensorFlow and defines all its own variables, so it shouldn't just have been caused by me being silly and accidentally reusing some otherwise defined b or having overwritten anything. This self-contained snippet still produced the error when just pasting and running it into its own jupyter notebook cell prior to restarting that particular kernel, but I guess that's not much to go on and the issue can be closed.
